parlai/agents/seq2seq/seq2seq.py

```diff
@@ -311,6 +311,12 @@ def __init__(self, opt, shared=None):
                           'changed.')
                 else:
                     self.optimizer.load_state_dict(states['optimizer'])
+                    if self.use_cuda:
+                        for state in self.optimizer.state.values():
+                            for k, v in state.items():
+                                if isinstance(v, torch.Tensor):
+                                    state[k] = v.cuda()
+
```

Comment:
maybe better to do states['optimizer'].pop('model') before `load_state_dict`? the optimizer should already have the model params loaded in from line 308

Reply:
Actually, looking at the generic code for [the Optimizer class](http://pytorch.org/docs/master/_modules/torch/optim/optimizer.html#Optimizer), the class actually needs the part of the state containing the old parameter list in order to correctly link together the tensors loaded from the saved optimizer to the tensors containing the model parameters (see the update_group method). So my impression is that it's preferable not to edit the state dict before we load it.

Reply:
@pemazare looks good to me. @alexholdenmiller ?

Thanks for fixing this!

Reply:
hm not yet I'm still concerned about iterating over a private field of the optimizer one sec